File: packages/kits/GeniusLocusKit/Tests/GeniusLocusKitTests/EncoderActivationTests.swift
// EncoderActivationTests.swift
//
// The `"encoder"` value of the `embedding_provider` manifest key and its
// failure contract: a device without the model directory (the default
// resolver) and a directory whose vocabulary hashes wrong both leave the
// estate with no encoder, log once, and never throw out of the open path;
// the Corpus ensemble stays the five-signal default. Plus the two companion
// manifest keys and their defaults.
//
// Failure modes: a factory error escaping `wireSubstores` (the provision
// would throw), the encoder being appended to the ensemble as if it were a
// provider, or a malformed `encoder_head` breaking the open instead of
// falling back to the default.
//
// Storage: SQLite so the manifest key written in one lifecycle survives the
// close + re-provision that runs `wireSubstores` again (same pattern as
// EmbeddingProviderConsumptionTests). `lifetime: .ephemeral` keeps Ed25519
// identities out of the real Keychain.

import Testing
import Foundation
import LocusKit
import CorpusKit
import CorpusKitProviders
import SynapseKit
import PersistenceKit
import PersistenceKitSQLite
@testable import GeniusLocusKit

private func scratchURL() -> URL {
    FileManager.default.temporaryDirectory
        .appendingPathComponent("glk-encoder-act-\(UUID().uuidString).sqlite3")
}

private func sqliteStorageAt(_ url: URL) throws -> any Storage {
    try SQLiteStorage(configuration: EstateConfiguration(
        estateID: UUID(),
        backend: .sqlite(url: url, busyTimeout: 5.0)))
}

private func glkParams(estateName: String) -> EstateProvisionParams {
    EstateProvisionParams(
        estateName: estateName,
        kind: .glk,
        zoomWindowLow: 1,
        zoomWindowHigh: 10,
        frameworkProfile: "KnowledgeWork",
        syncMode: .none,
        lifetime: .ephemeral)
}

/// Resolver that points every model id at one fixed directory.
private struct FixedDirectoryResolver: ModelDirectoryResolving {
    let directory: URL
    func encoderModelDirectory(for modelID: String) -> URL? { directory }
}

/// Resolver that records every model id requested and returns nil (no
/// model directory on this device). Used to prove activation read the
/// seeded row id rather than the floor model id.
private final class SpyResolver: ModelDirectoryResolving, @unchecked Sendable {
    private let lock = NSLock()
    private var _requested: [String] = []

    var requested: [String] {
        lock.lock(); defer { lock.unlock() }
        return _requested
    }

    func encoderModelDirectory(for modelID: String) -> URL? {
        lock.lock(); defer { lock.unlock() }
        _requested.append(modelID)
        return nil
    }
}

@Suite("Encoder activation — constants")
struct EncoderActivationConstantTests {

    @Test("provider id and manifest keys are the wire strings")
    func wireStrings() {
        #expect(GeniusLocusKit.encoderProviderID == "encoder")
        #expect(GeniusLocusKit.encoderHeadMetaKey == "encoder_head")
        #expect(GeniusLocusKit.encoderBatchMetaKey == "encoder_batch")
        #expect(GeniusLocusKit.encoderHeadMetaKey != GeniusLocusKit.embeddingProviderMetaKey)
    }

    @Test("defaults are head 30 and batch 64 off iOS")
    func defaults() {
        #expect(GeniusLocusKit.defaultEncoderHead == 30)
#if os(iOS)
        #expect(GeniusLocusKit.defaultEncoderBatch == 16)
#else
        #expect(GeniusLocusKit.defaultEncoderBatch == 64)
#endif
    }
}

@Suite("Encoder activation — failure contract and manifest keys", .serialized)
struct EncoderActivationEstateTests {

    @Test("provision writes the default encoder key; an estate that names a provider keeps it")
    func provisionDefaultEncoder() async throws {
        let kit = GeniusLocusKit()
        let url = scratchURL()
        let owner = OwnerCredentials(ownerIdentifier: "encoder-act-default")
        let params = glkParams(estateName: "EncoderDefaultEstate")

        // A fresh provision is born with the encoder as its recall stage.
        let handle = try await kit.provision(storage: try sqliteStorageAt(url), owner: owner, params: params)
        #expect(try await kit.provisionedEmbeddingProvider(for: handle) == GeniusLocusKit.encoderProviderID)
        // Idempotent: a second call writes nothing.
        #expect(try await kit.provisionDefaultEncoderIfAbsent(for: handle) == false)
        // An estate that names another provider is never overwritten.
        try await kit.provisionEmbeddingProvider("apple-nl-v1", for: handle)
        #expect(try await kit.provisionDefaultEncoderIfAbsent(for: handle) == false)
        #expect(try await kit.provisionedEmbeddingProvider(for: handle) == "apple-nl-v1")
        // A cleared key is absent again and the default returns.
        try await kit.provisionEmbeddingProvider("", for: handle)
        #expect(try await kit.provisionDefaultEncoderIfAbsent(for: handle) == true)
        #expect(try await kit.provisionedEmbeddingProvider(for: handle) == GeniusLocusKit.encoderProviderID)
        try await kit.close(handle)
    }

    @Test("encoder provisioned without a model directory: no encoder, ensemble untouched, no throw")
    func noModelDirectory() async throws {
        let kit = GeniusLocusKit()
        let url = scratchURL()
        let owner = OwnerCredentials(ownerIdentifier: "encoder-act-nodir")
        let params = glkParams(estateName: "EncoderNoDirEstate")

        let handle1 = try await kit.provision(storage: try sqliteStorageAt(url), owner: owner, params: params)
        try await kit.provisionEmbeddingProvider(GeniusLocusKit.encoderProviderID, for: handle1)
        try await kit.close(handle1)

        // Re-provision runs wireSubstores → applyProvisionedEmbeddingProvider
        // → activateSpanEncoder with the default (nil) resolver.
        let handle2 = try await kit.provision(storage: try sqliteStorageAt(url), owner: owner, params: params)
        #expect(await kit.registeredSpanEncoder(for: handle2) == nil)
        let corpus = try #require(await kit.corpusKits[handle2])
        let modelIDs = await corpus.providerGenerations().map(\.modelID)
        // The ensemble is the default dense ensemble exactly — Random Indexing
        // plus the LSA whole-record float provider (GENIUSLOCUSKIT_SPEC 3.32.0)
        // — and the encoder is never a member of it.
        #expect(modelIDs == ["random-indexing-v1", "lsa-v1"],
                "the encoder is never an ensemble member, got \(modelIDs)")
        try await kit.close(handle2)
    }

    @Test("encoder provisioned with a wrong-vocab directory: factory error stays inside the lifecycle")
    func hashMismatchDirectory() async throws {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("glk-encoder-act-model-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        try Data("[PAD]\n[UNK]\n[CLS]\n[SEP]\nhello\n".utf8).write(to: dir.appendingPathComponent("vocab.txt"))

        let kit = GeniusLocusKit()
        await kit.setModelDirectoryResolver(FixedDirectoryResolver(directory: dir))
        let url = scratchURL()
        let owner = OwnerCredentials(ownerIdentifier: "encoder-act-hash")
        let params = glkParams(estateName: "EncoderHashEstate")

        let handle1 = try await kit.provision(storage: try sqliteStorageAt(url), owner: owner, params: params)
        try await kit.provisionEmbeddingProvider(GeniusLocusKit.encoderProviderID, for: handle1)
        try await kit.close(handle1)

        // The resolver returns the directory; the factory throws
        // tokenizerMismatch; the lifecycle swallows it (one log line).
        let handle2 = try await kit.provision(storage: try sqliteStorageAt(url), owner: owner, params: params)
        #expect(await kit.registeredSpanEncoder(for: handle2) == nil)
        try await kit.close(handle2)
    }

    @Test("provision seeds the active encoder_models row and activation reads it")
    func provisionSeedsTheActiveEncoderRowAndActivatesUnderIt() async throws {
        // SpyResolver records which model ids activation requests; returns nil
        // so no model loads (tests the seeding path, not the encoder load path).
        let spy = SpyResolver()
        let kit = GeniusLocusKit()
        await kit.setModelDirectoryResolver(spy)
        let url = scratchURL()
        let owner = OwnerCredentials(ownerIdentifier: "encoder-act-seed")
        let storage = try sqliteStorageAt(url)

        let handle = try await kit.provision(storage: storage, owner: owner,
                                             params: glkParams(estateName: "EncoderSeedEstate"))

        // The seam seeds the bundled row as the active entry at open.
        let registry = EncoderModelStore(storage: storage)
        let activeRow = try await registry.active()
        #expect(activeRow?.modelID == EncoderModelSeed.modelID)
        #expect(activeRow?.isActive == true)

        // Activation asked the resolver for the seeded row's id, not the
        // floor model (minilm-l6-v2-w60), proving it read the seeded row.
        #expect(spy.requested == [EncoderModelSeed.modelID])

        // Spy returns nil: no model directory, so no encoder and no rerank stage.
        #expect(await kit.isSpanRerankRegistered(for: handle) == false)
        #expect(await kit.registeredSpanEncoder(for: handle) == nil)

        // Capture one drawer; assert the VectorStore has no span rows yet
        // (the span-encode standing signal has not run).
        let frame = CaptureFrame(content: "encoder seed test content", channel: .typed,
                                 room: "encoder-seed-tests", latticeAnchor: .udc("000"),
                                 addedBy: "encoder-seed-tests", embeddingModelID: "test-v1")
        let drawer = try await kit.capture(handle, frame)
        let vectors = try #require(await kit.registeredVectorStore(for: handle))
        let spanRows = try await vectors.spanVectors(itemIDs: [drawer.id],
                                                     modelID: EncoderModelSeed.modelID)
        #expect(spanRows.isEmpty)

        // Seeding is idempotent: the active row already exists.
        let wasSeeded = try await kit.seedDefaultEncoderModelIfAbsent(for: handle)
        #expect(wasSeeded == false)
        #expect(try await registry.all().count == 1)

        try await kit.close(handle)
    }

    @Test("serve open path seeds the encoder row through wireGLKSubstores")
    func serveOpenPathSeedsThroughWireGLKSubstores() async throws {
        let url = scratchURL()
        let owner = OwnerCredentials(ownerIdentifier: "encoder-seed-serve")

        // LocusOnly provision so the encoder key is NOT written (the provision
        // path gates on params.kind != .locusOnly). This mimics the first-run
        // state of an estate provisioned before the encoder was configured.
        let kit1 = GeniusLocusKit()
        let locusParams = EstateProvisionParams(
            estateName: "EncoderServeSeedEstate",
            kind: .locusOnly,
            zoomWindowLow: 1, zoomWindowHigh: 10,
            frameworkProfile: "KnowledgeWork",
            syncMode: .none, lifetime: .ephemeral)
        let h1 = try await kit1.provision(storage: try sqliteStorageAt(url), owner: owner,
                                          params: locusParams)
        try await kit1.close(h1)

        // Serve open path: open + provisionDefaultEncoderIfAbsent + wireGLKSubstores.
        let kit2 = GeniusLocusKit()
        let storage2 = try sqliteStorageAt(url)
        let handle = try await kit2.open(storage: storage2, owner: owner,
                                         identityKeyStore: InMemoryEstateIdentityKeyStore())

        // Write the encoder key (ServeCommand does this on first run).
        let keyWritten = try await kit2.provisionDefaultEncoderIfAbsent(for: handle)
        #expect(keyWritten == true)

        // The registry is empty before wiring: no row has been seeded yet.
        let registry = EncoderModelStore(storage: storage2)
        #expect(try await registry.active() == nil)

        // Wire substores: activateSpanEncoderIfProvisioned runs and seeds the row.
        try await kit2.wireGLKSubstores(for: handle, backingStorage: storage2)

        // After wiring the bundled row is present as the active entry.
        let activeRow = try await registry.active()
        #expect(activeRow?.modelID == EncoderModelSeed.modelID)

        try await kit2.close(handle)
    }

    /// Runs only where the Arctic CoreML directory exists (`MOOT_ENCODER_MODEL_DIR`
    /// names it): the seeded row loads, the rerank stage registers on the fresh
    /// estate, and a captured drawer still has no span rows until the span-encode
    /// signal runs. Skipped elsewhere; the report records whether it ran.
    @Test("a real model directory registers the rerank stage on a fresh estate",
          .enabled(if: ProcessInfo.processInfo.environment["MOOT_ENCODER_MODEL_DIR"] != nil))
    func realModelRegistersTheRerankStage() async throws {
        let path = try #require(ProcessInfo.processInfo.environment["MOOT_ENCODER_MODEL_DIR"])
        let kit = GeniusLocusKit()
        await kit.setModelDirectoryResolver(FixedDirectoryResolver(directory: URL(fileURLWithPath: path, isDirectory: true)))
        let storage = try sqliteStorageAt(scratchURL())
        let owner = OwnerCredentials(ownerIdentifier: "encoder-act-real")
        let handle = try await kit.provision(storage: storage, owner: owner,
                                             params: glkParams(estateName: "EncoderRealModelEstate"))
        #expect(try await EncoderModelStore(storage: storage).active()?.modelID == EncoderModelSeed.modelID)
        #expect(await kit.isSpanRerankRegistered(for: handle) == true)
        let frame = CaptureFrame(content: "encoder real model test content", channel: .typed,
                                 room: "encoder-real-tests", latticeAnchor: .udc("000"),
                                 addedBy: "encoder-real-tests", embeddingModelID: "test-v1")
        let drawer = try await kit.capture(handle, frame)
        let vectors = try #require(await kit.registeredVectorStore(for: handle))
        #expect(try await vectors.spanVectors(itemIDs: [drawer.id], modelID: EncoderModelSeed.modelID).isEmpty)
        try await kit.close(handle)
    }

    @Test("encoder_head / encoder_batch round-trip, malformed values fall back to defaults")
    func manifestKeys() async throws {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "encoder-act-keys")
        let handle = try await kit.provision(
            storage: try sqliteStorageAt(scratchURL()), owner: owner,
            params: glkParams(estateName: "EncoderKeysEstate"))

        #expect(await kit.provisionedEncoderHead(for: handle) == GeniusLocusKit.defaultEncoderHead)
        #expect(await kit.provisionedEncoderBatch(for: handle) == GeniusLocusKit.defaultEncoderBatch)

        try await kit.provisionEncoderHead(45, for: handle)
        try await kit.provisionEncoderBatch(8, for: handle)
        #expect(await kit.provisionedEncoderHead(for: handle) == 45)
        #expect(await kit.provisionedEncoderBatch(for: handle) == 8)

        // A malformed or non-positive value never breaks a read: default.
        let estate = try await kit.estate(for: handle)
        try await estate.setMeta(key: GeniusLocusKit.encoderHeadMetaKey, value: "thirty")
        try await estate.setMeta(key: GeniusLocusKit.encoderBatchMetaKey, value: "0")
        #expect(await kit.provisionedEncoderHead(for: handle) == GeniusLocusKit.defaultEncoderHead)
        #expect(await kit.provisionedEncoderBatch(for: handle) == GeniusLocusKit.defaultEncoderBatch)
        try await kit.close(handle)
    }
}

File: packages/kits/GeniusLocusKit/rust/tests/encoder_activation_tests.rs
//! The `"encoder"` value of the `embedding_provider` manifest key and its
//! failure contract on the Rust coordinator: the default (nil) resolver and
//! a directory whose vocabulary hashes wrong both leave the estate with no
//! encoder and never panic or error out of `apply_provisioned_embedding_provider`;
//! plus the two companion manifest keys and their defaults.
//!
//! Failure modes: a factory error escaping the apply path, or a malformed
//! `encoder_head` value breaking the read instead of falling back.
//!
//! The seeding tests (`encoder_activation_seeds_the_default_encoder_row` and
//! `wire_glk_substores_seeds_through_the_open_path`) verify Bob's ruling
//! 2026-09-04: upgrade never creates content; seeding belongs to provision and
//! serve. `activate_span_encoder` seeds the bundled row before reading the
//! registry so an estate is encoder-active from its first open. Twin of
//! Swift `EncoderActivationTests.provisionSeedsTheActiveEncoderRowAndActivatesUnderIt`
//! and `serveOpenPathSeedsThroughWireGLKSubstores`.

use std::collections::BTreeMap;
use std::path::PathBuf;
use std::sync::Arc;

use genius_locus_kit::{EstateCoordinator, ModelDirectoryResolving};
use locus_kit::{
    drawer_store::DrawerStore, drawer_store_inmemory::InMemoryDrawerStore,
    estate_types::OwnerCredentials,
};
use persistence_kit::{storage::Storage, TypedValue};

const NOW: i64 = 1_700_000_000;

fn open_one() -> (EstateCoordinator, genius_locus_kit::EstateHandle) {
    let mut coord = EstateCoordinator::new();
    let store: Arc<dyn DrawerStore> = Arc::new(InMemoryDrawerStore::new(NOW, None).unwrap());
    let handle = coord
        .open(store, OwnerCredentials::new("encoder-act-test"), 0, i64::MAX)
        .expect("open must succeed");
    (coord, handle)
}

/// Resolver that points every model id at one fixed directory.
struct FixedDirectoryResolver(PathBuf);

impl ModelDirectoryResolving for FixedDirectoryResolver {
    fn model_dir_for(&self, _model_id: &str) -> Option<PathBuf> {
        Some(self.0.clone())
    }
}

#[test]
fn wire_strings_and_defaults() {
    assert_eq!(EstateCoordinator::ENCODER_PROVIDER_ID, "encoder");
    assert_eq!(EstateCoordinator::ENCODER_HEAD_META_KEY, "encoder_head");
    assert_eq!(EstateCoordinator::ENCODER_BATCH_META_KEY, "encoder_batch");
    assert_ne!(
        EstateCoordinator::ENCODER_HEAD_META_KEY,
        EstateCoordinator::EMBEDDING_PROVIDER_META_KEY
    );
    assert_eq!(EstateCoordinator::DEFAULT_ENCODER_HEAD, 30);
    assert_eq!(EstateCoordinator::DEFAULT_ENCODER_BATCH, 64);
}

#[test]
fn encoder_without_model_directory_registers_nothing_and_does_not_panic() {
    let (mut coord, handle) = open_one();
    coord
        .provision_embedding_provider(&handle, EstateCoordinator::ENCODER_PROVIDER_ID)
        .expect("provision");
    // Default resolver → no directory → one stderr line, no encoder.
    coord.apply_provisioned_embedding_provider(&handle);
    assert!(coord.registered_span_encoder(&handle).is_none());
}

#[test]
fn encoder_with_wrong_vocab_directory_registers_nothing_and_does_not_panic() {
    let dir = std::env::temp_dir().join(format!("glk-encoder-act-{}", std::process::id()));
    let _ = std::fs::remove_dir_all(&dir);
    std::fs::create_dir_all(&dir).unwrap();
    std::fs::write(dir.join("vocab.txt"), b"[PAD]\n[UNK]\n[CLS]\n[SEP]\nhello\n").unwrap();

    let (mut coord, handle) = open_one();
    coord.set_model_directory_resolver(Box::new(FixedDirectoryResolver(dir.clone())));
    coord
        .provision_embedding_provider(&handle, EstateCoordinator::ENCODER_PROVIDER_ID)
        .expect("provision");
    // Resolver answers; the factory reports TokenizerMismatch; the
    // coordinator swallows it (one stderr line).
    coord.apply_provisioned_embedding_provider(&handle);
    let _ = std::fs::remove_dir_all(&dir);
    assert!(coord.registered_span_encoder(&handle).is_none());
}

#[test]
fn encoder_head_and_batch_round_trip_with_defaults_on_malformed_values() {
    let (coord, handle) = open_one();
    assert_eq!(coord.provisioned_encoder_head(&handle), EstateCoordinator::DEFAULT_ENCODER_HEAD);
    assert_eq!(coord.provisioned_encoder_batch(&handle), EstateCoordinator::DEFAULT_ENCODER_BATCH);

    coord.provision_encoder_head(&handle, 45).expect("head");
    coord.provision_encoder_batch(&handle, 8).expect("batch");
    assert_eq!(coord.provisioned_encoder_head(&handle), 45);
    assert_eq!(coord.provisioned_encoder_batch(&handle), 8);

    // Malformed / non-positive values never break a read: default.
    let estate = coord.estate_for(&handle).expect("estate");
    estate.set_meta(EstateCoordinator::ENCODER_HEAD_META_KEY, "thirty").unwrap();
    estate.set_meta(EstateCoordinator::ENCODER_BATCH_META_KEY, "0").unwrap();
    assert_eq!(coord.provisioned_encoder_head(&handle), EstateCoordinator::DEFAULT_ENCODER_HEAD);
    assert_eq!(coord.provisioned_encoder_batch(&handle), EstateCoordinator::DEFAULT_ENCODER_BATCH);
}

#[test]
fn close_drops_the_span_encoder_registration_slot() {
    // No encoder can load here, so this pins the close path indirectly: a
    // closed handle answers None and does not panic.
    let (mut coord, handle) = open_one();
    coord.close(&handle).expect("close");
    assert!(coord.registered_span_encoder(&handle).is_none());
}

/// `provision_default_encoder_if_absent` writes `"encoder"` only when the
/// estate names no provider, and never overwrites a named one. Swift twin:
/// EncoderActivationTests.provisionDefaultEncoder.
#[test]
fn default_encoder_provisioning_writes_once_and_never_overwrites() {
    let (coord, handle) = open_one();
    assert_eq!(coord.provisioned_embedding_provider(&handle).unwrap(), None);
    assert!(coord.provision_default_encoder_if_absent(&handle).unwrap());
    assert_eq!(
        coord.provisioned_embedding_provider(&handle).unwrap().as_deref(),
        Some(EstateCoordinator::ENCODER_PROVIDER_ID)
    );
    // Idempotent: a second call writes nothing.
    assert!(!coord.provision_default_encoder_if_absent(&handle).unwrap());
    // A named provider is never overwritten.
    coord.provision_embedding_provider(&handle, "apple-nl-v1").unwrap();
    assert!(!coord.provision_default_encoder_if_absent(&handle).unwrap());
    assert_eq!(
        coord.provisioned_embedding_provider(&handle).unwrap().as_deref(),
        Some("apple-nl-v1")
    );
    // A cleared key is absent again and the default returns.
    coord.provision_embedding_provider(&handle, "").unwrap();
    assert!(coord.provision_default_encoder_if_absent(&handle).unwrap());
    assert_eq!(
        coord.provisioned_embedding_provider(&handle).unwrap().as_deref(),
        Some(EstateCoordinator::ENCODER_PROVIDER_ID)
    );
}

// ─── Seeding tests (Part B, ruling 2026-09-04) ────────────────────────────────

/// Records every model ID the resolver is queried for; returns None so no
/// encoder loads. Shared via Arc so the test can inspect captured calls after
/// the coordinator owns the Box<dyn ModelDirectoryResolving>.
struct SpyResolver(std::sync::Arc<std::sync::Mutex<Vec<String>>>);

impl ModelDirectoryResolving for SpyResolver {
    fn model_dir_for(&self, model_id: &str) -> Option<PathBuf> {
        self.0.lock().unwrap().push(model_id.to_string());
        None
    }
}

/// Returns the coordinator, handle, AND the DrawerStore Arc so tests can
/// reach the estate's underlying storage for encoder_model_store reads.
fn open_one_with_store() -> (EstateCoordinator, genius_locus_kit::EstateHandle, Arc<dyn DrawerStore>) {
    let mut coord = EstateCoordinator::new();
    let store: Arc<dyn DrawerStore> = Arc::new(InMemoryDrawerStore::new(NOW, None).unwrap());
    let store_ref = Arc::clone(&store);
    let handle = coord
        .open(store, OwnerCredentials::new("encoder-seed-test"), 0, i64::MAX)
        .expect("open must succeed");
    (coord, handle, store_ref)
}

/// `activate_span_encoder` seeds the bundled encoder row before reading the
/// registry when no active row exists, so an estate provisioned with the
/// encoder key is encoder-active from its first open. The resolver spy
/// confirms the seeded MODEL_ID was read, not the floor model.
///
/// Twin of Swift `EncoderActivationTests.provisionSeedsTheActiveEncoderRowAndActivatesUnderIt`.
#[test]
fn encoder_activation_seeds_the_default_encoder_row() {
    use corpus_kit_providers::EncoderModelSeed;
    use locus_kit::encoder_model_store::EncoderModelStore;

    let spy_calls = std::sync::Arc::new(std::sync::Mutex::new(Vec::<String>::new()));
    let (mut coord, handle, store) = open_one_with_store();
    coord.set_model_directory_resolver(Box::new(SpyResolver(std::sync::Arc::clone(&spy_calls))));
    coord
        .provision_embedding_provider(&handle, EstateCoordinator::ENCODER_PROVIDER_ID)
        .expect("provision");

    // Activation seeds first: after this call the active row must exist.
    coord.apply_provisioned_embedding_provider(&handle);

    let storage = store.storage().expect("InMemoryDrawerStore must expose its storage");
    let registry = EncoderModelStore::new(storage);
    let row = registry
        .active()
        .expect("active() must not fail")
        .expect("active row must exist after activation seeds it");
    assert_eq!(row.model_id, EncoderModelSeed::MODEL_ID);
    assert!(row.is_active);
    // The resolver saw MODEL_ID, proving activation read the seeded row and
    // not the floor model (minilm-l6-v2-w60).
    let seen = spy_calls.lock().unwrap();
    assert_eq!(seen.as_slice(), &[EncoderModelSeed::MODEL_ID]);
    drop(seen);
    // No model directory in tests: encoder and rerank stage are not registered.
    assert!(coord.registered_span_encoder(&handle).is_none());
    assert!(!coord.is_span_rerank_registered(&handle));
    // seed_default_encoder_model_if_absent is idempotent.
    let again = coord
        .seed_default_encoder_model_if_absent(&handle)
        .expect("seed must not fail");
    assert!(!again, "seed must be idempotent when a row already exists");
    assert_eq!(registry.all().expect("all() must not fail").len(), 1);
}

/// `wire_glk_substores` seeds the active encoder row through
/// `activate_span_encoder`, so the serve-open path activates the encoder from
/// the first open of a provisioned estate.
///
/// Twin of Swift `EncoderActivationTests.serveOpenPathSeedsThroughWireGLKSubstores`.
#[test]
fn wire_glk_substores_seeds_through_the_open_path() {
    use corpus_kit_providers::{default_ensemble, EncoderModelSeed};
    use genius_locus_kit::estate_format::{EstateFormatStore, EstateFormatVersion};
    use locus_kit::drawer_operational::CaptureChannel;
    use locus_kit::encoder_model_store::EncoderModelStore;
    use locus_kit::estate_types::LatticeAnchor;
    use locus_kit::frames::CaptureFrame;
    use persistence_kit::inmemory::InMemoryStorage;
    use uuid::Uuid;

    let (mut coord, handle, store) = open_one_with_store();
    coord.provision_default_encoder_if_absent(&handle).expect("provision encoder key");

    let storage = store.storage().expect("storage must be accessible");
    let registry = EncoderModelStore::new(Arc::clone(&storage));
    // Pre-wire: no active row; seeding happens inside activate_span_encoder.
    assert!(
        registry.active().expect("active() must not fail").is_none(),
        "no row before wire_glk_substores"
    );

    // A fresh InMemoryStorage stamped at the current estate format, mirroring
    // wire_inmemory_semantic_recall in AriaMcpKit/rust/src/estate_registry.rs.
    let backing: Arc<dyn persistence_kit::storage::Storage> =
        Arc::new(InMemoryStorage::with_estate(Uuid::new_v4()));
    EstateFormatStore::new(Arc::clone(&backing))
        .stamp(EstateFormatVersion::CURRENT, NOW)
        .expect("format stamp must succeed");

    coord
        .wire_glk_substores(&handle, backing, default_ensemble(), NOW)
        .expect("wire_glk_substores must succeed");

    // Post-wire: the active row must be seeded with MODEL_ID.
    let row = registry
        .active()
        .expect("active() must not fail")
        .expect("active row must be seeded by wire_glk_substores");
    assert_eq!(row.model_id, EncoderModelSeed::MODEL_ID);

    // Span rows are empty until the span-encode signal runs.
    let frame = CaptureFrame::new(
        "hello world",
        CaptureChannel::Typed,
        "inbox",
        LatticeAnchor::udc("0"),
        "test",
        EncoderModelSeed::MODEL_ID,
    );
    let drawer = coord.capture(&handle, frame, NOW).expect("capture must succeed");
    let span_rows = coord
        .vector_store_for(&handle)
        .expect("vector store must be registered after wire_glk_substores")
        .span_vectors(&[drawer.id.as_str()], EncoderModelSeed::MODEL_ID)
        .expect("span_vectors must not fail");
    assert!(span_rows.is_empty(), "span rows are empty until the signal drains them");
}

fn rows(storage: &Arc<dyn Storage>, table: &str) -> Vec<BTreeMap<String, TypedValue>> {
    storage.row_store().query(table, None, &[], None, None)
        .unwrap_or_else(|error| panic!("read {table}: {error:?}"))
        .into_iter()
        .map(|row| row.values)
        .collect()
}

#[test]
fn readonly_glk_wiring_preserves_stale_receipts_and_keeps_query_tiers() {
    use corpus_kit_providers::default_ensemble;
    use genius_locus_kit::estate_format::{EstateFormatStore, EstateFormatVersion};
    use persistence_kit::inmemory::InMemoryStorage;
    use uuid::Uuid;

    let (mut coord, handle, _drawer) = open_one_with_store();
    let backing: Arc<dyn Storage> = Arc::new(InMemoryStorage::with_estate(Uuid::new_v4()));
    EstateFormatStore::new(Arc::clone(&backing))
        .stamp(EstateFormatVersion::CURRENT, NOW)
        .expect("current format");
    coord.wire_glk_substores(&handle, Arc::clone(&backing), default_ensemble(), NOW)
        .expect("live preparation");

    let mut configuration = BTreeMap::new();
    configuration.insert("singleton_id".to_string(), TypedValue::Int(1));
    configuration.insert("generation_token".to_string(), TypedValue::Text("stale-ri-fingerprint".to_string()));
    configuration.insert("updated_at".to_string(), TypedValue::Timestamp(17));
    backing.row_store().upsert("corpus_provider_configuration", configuration, &["singleton_id".to_string()])
        .expect("stale configuration");
    for mut claim in rows(&backing, "vector_rep_claims") {
        claim.insert("claimed_at".to_string(), TypedValue::Timestamp(19));
        backing.row_store().upsert("vector_rep_claims", claim, &[
            "model_id".to_string(), "model_version".to_string(),
            "vector_index".to_string(), "consumer".to_string(),
        ]).expect("stale claim");
    }
    let expected_configuration = rows(&backing, "corpus_provider_configuration");
    let expected_claims = rows(&backing, "vector_rep_claims");
    assert!(!expected_claims.is_empty(), "live preparation must create vector claims");

    coord.wire_glk_substores_readonly(&handle, Arc::clone(&backing), default_ensemble(), NOW + 1_000)
        .expect("read-preserving wire");
    assert!(coord.has_corpus(&handle), "readonly wire keeps the corpus query tier");
    assert!(coord.has_vector_store(&handle), "readonly wire keeps the vector/strict source tier");
    assert_eq!(rows(&backing, "corpus_provider_configuration"), expected_configuration,
        "readonly wire must not reconcile a stale provider receipt");
    assert_eq!(rows(&backing, "vector_rep_claims"), expected_claims,
        "readonly wire must not refresh claim timestamps");
}

/// When a real Arctic CoreML model directory is available, the rerank stage is
/// registered after activation seeds and activates the encoder.
///
/// Requires `MOOT_ENCODER_MODEL_DIR` set to a directory containing the Arctic
/// vocab file and weights. Run with `--features encoder` to compile the
/// model-loading path.
///
/// The estate must be wired through `wire_glk_substores` so the VectorStore
/// is registered before `apply_provisioned_embedding_provider` runs.
/// `activate_span_encoder` only registers the rerank stage when a VectorStore
/// is present — without the wire step the assertion would always fail even
/// when the real model loads. Mirrors Swift `kit.provision()`.
#[test]
#[ignore = "needs MOOT_ENCODER_MODEL_DIR and --features encoder"]
fn real_model_registers_the_rerank_stage() {
    use corpus_kit_providers::default_ensemble;
    use genius_locus_kit::estate_format::{EstateFormatStore, EstateFormatVersion};
    use persistence_kit::inmemory::InMemoryStorage;
    use uuid::Uuid;

    let dir = std::env::var("MOOT_ENCODER_MODEL_DIR")
        .expect("MOOT_ENCODER_MODEL_DIR must point at the Arctic model directory");
    let (mut coord, handle, _store) = open_one_with_store();
    coord.set_model_directory_resolver(Box::new(FixedDirectoryResolver(PathBuf::from(&dir))));
    coord
        .provision_embedding_provider(&handle, EstateCoordinator::ENCODER_PROVIDER_ID)
        .expect("provision");
    // Wire GLK substores so the VectorStore is registered before
    // apply_provisioned_embedding_provider runs (called inside wire_glk_substores).
    // A fresh InMemoryStorage stamped at the current estate format mirrors the
    // wire_inmemory_semantic_recall pattern in AriaMcpKit/rust/src/estate_registry.rs.
    let backing: Arc<dyn persistence_kit::storage::Storage> =
        Arc::new(InMemoryStorage::with_estate(Uuid::new_v4()));
    EstateFormatStore::new(Arc::clone(&backing))
        .stamp(EstateFormatVersion::CURRENT, NOW)
        .expect("format stamp must succeed");
    coord
        .wire_glk_substores(&handle, backing, default_ensemble(), NOW)
        .expect("wire_glk_substores must succeed");
    assert!(
        coord.is_span_rerank_registered(&handle),
        "rerank stage must be registered when a real model directory is provided"
    );
}
